tests/unit/components/playback/controls/AddCustomer.spec.js

```diff
@@ -176,7 +178,8 @@ describe("AddCustomer.enter_uuidbase57", () => {
       "nickname-id",
       "validate_nickname",
     ],
-    ["C", "minimum one character C", "nickname-id", "validate_nickname", ""],
+    ["C", "minimum one character C", "nickname-id", "validate_nickname"],
+    ["", "<empty>", "nickname-id", "validate_nickname"],
```

Comment:
As I mentioned in Asana, whether a field is optional or required is a specific decision that the parent component (e.g. AddCustomer) makes in its internal business logic.   This is independent of the TextValidation logic--for example currently all of the places we ask for an Alphanumeric ID are required, but we could have an input in the future where we want that to be optional. The TextValidation function let's us know whether a string is a valid Base57 UUID or not, but it is the parent component's business logic that decides whether an empty string should be considered valid or not.

So in the case of a blank input, it will likely require a different test structure. Because for this we want to verify not that the message in the DOM matches the returned value of TextValidation, but specifically whether the value in the DOM is an empty string (if the business logic of the component says that this field is optional) or that it matches the specific text the parent wants displayed if the field is required an the textbox is empty (which might be just something simple like "This field is required" or something more complex like "An API Key is required").

Please take this comment into account for the tests for all 4 new components you made in this Pull Request.

Reply:
As of my understanding we are having only two components and one field which is optional that is "API Key(Optional)". 
a) AddCustomer 
b) EditCustomer

For now we have a testcase which verifies on the feedback obtained from the validation
which determines by a function "enable_save_button" which validates the "Optional" or non entered or field, we can verify in testcase if still the save button is available by the
color set, a similar one can be done for the EditCustomer.

describe("AddCustomer.enable_save_button"
[
      "5FY8KwTsQaUJ2KzHJGetfE",
      "",
      "Experiment anemia -1",
      "",
      "",
      "",
      "color: rgb(255, 255, 255);",
    ],

For the other two components:-
a)AddUser 
b)EditUser

The fields are a must required without which the forward functional "Save ID" is not active so there is no "optional" and so nothing needs to be verified
and its covered.

Hope this covers the optional testing.. for the components at present in hand its just 
in two components AddCustomer and Edit Customer

Please let me does the above inclusion of testcase is sufficient ?

Reply:
The logic of what makes a global button in a component (e.g. the `Save ID` button) active or disabled should be tested separately from the individual logic of each TextInput field.  If we decide to change the rules that make `Save ID` clickable, we don't want changing those tests to mean we lose coverage of how an individual TextInput field should operate---for example, we could decide in the future that we want the `Save ID` button to always be enabled, and then display a message if someone tries to click it when not all the fields are enabled.  In general though, these are two separate pieces of business logic and we want to test them separately.

Your tests for the `Save ID` button are good and cover a variety of situations and therefore ensure that it operates the way we expect.

When utilizing TextInput widgets within a component we are building, we want to make sure they are wired up correctly by testing the following:

- The correct TextValidator function is called when the field is updated

- When two distinct invalid entries are entered, the error message in the DOM displays the message returned by the TextValidator

- When two distinct valid entries are entered, the error message in the DOM is blank (the widget is displaying as valid)

- When the field is empty, it behaves as the business logic of the parent intends: if the field is optional then the error message in the DOM should be blank (the widget is displaying as valid), and if the field is required then the error message in the DOM should display the exact text that the parent component intends (which could vary based on the business logic and should not be directly reliant on the value returned by the TextValidator function----the TextValidator function messages should be universal if someone types something that is wrong, but there are many situations where a parent component may want to vary the message if the field is blank but still required)


Only the last point needs to be implemented still

Reply:
Agreed, but I don't know how to convert this into a test scenario. Could we discuss in the daily sync up.

Reply:
yes, we can discuss today